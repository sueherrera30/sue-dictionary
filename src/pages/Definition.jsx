import React, { Fragment } from "react";
import { useParams, useNavigate } from 'react-router-dom';
import { Tooltip, IconButton, CircularProgress } from '@mui/material';
import AddIcon from '@mui/icons-material/Add';
import DoneOutlineIcon from '@mui/icons-material/DoneOutline';

import { BsBookmarkHeart as BookmarkHeart } from "react-icons/bs";
import { MdOutlineArrowBack as Back } from "react-icons/md";
import { GiSpeaker as Speaker } from "react-icons/gi";

import useFetch from '../utils/hooks/useFetch';
import ErrorComponent from '../components/Error';
import DefinitionBox from '../components/DefinitionBox';
import sharedStyles from './styles/home.module.css';
import styles from './styles/definition.module.css';

const Definition = ({ favorites, setFavorites }) => {
    const { word } = useParams();

    let navigate = useNavigate();
    const { definitions, loading, error, audio } = useFetch(word);

    console.log('error in definition component', error);
    const handleAddFavorites = () => {
        setFavorites(prevFavs => [...prevFavs, word])
        if(favorites.includes(word)) return;
    }
    const removeFavorites = () => {
        const filteredFavs = favorites.filter(item => item !== word);
        setFavorites(filteredFavs)
    } 
    if(loading) return <div className={styles.loadingContainer}><CircularProgress className={styles.loading} color="secondary" /></div>
    if(error !== null) return <ErrorComponent error={error}/>

    return (
        <section className={sharedStyles.mainContainer}>
            <div className={`${sharedStyles.container} ${styles.container}`}>
                {
                    !definitions 
                    ? null
                    : <>
                    <div className={styles.iconsContainer}>
                        <Back data-testid="back-button" className={styles.icons} onClick={() => navigate('/')} />
                        <div className={styles.favsIcons}>
                            {
                                favorites && favorites.includes(word)
                                ? <IconButton data-testid="stick-icon" onClick={removeFavorites}> <DoneOutlineIcon className={styles.stick} /></IconButton>
                                : <Tooltip title="add to favs" arrow  enterTouchDelay={0}>
                                    <IconButton  data-testid="add-icon"  onClick={handleAddFavorites}> <AddIcon className={styles.add}  /></IconButton>
                                </Tooltip>
                            }
                            <Tooltip title="Go to your favourites" arrow  enterTouchDelay={0}>
                                <div><BookmarkHeart  data-testid="favorites-button"  className={styles.icons} onClick={() => navigate('/favorites')}/></div>
                            </Tooltip>
                        </div>
                    </div>
                    <div className={styles.wordContainer}>
                        <div>
                        <p className={styles.wordTitle} data-testid="word">{word}</p>
                        </div>
                        {audio ? <Speaker onClick={() => audio.play()} className={`${styles.icons} ${styles.soundIcon}`} /> : null}
                    </div>
                    <div className={styles.definitionsContainer}>
                        {
                            definitions.map((definition, idx) => 
                                <Fragment key={idx}>
                                    {
                                        definition.meanings.map((meaning, idx) =>
                                            <DefinitionBox
                                                key={idx}
                                                meanings={meaning.definitions}
                                                partOfSpeech={meaning.partOfSpeech}
                                                phonetic={definition.phonetic}
                                            />
                                        )
                                    }
                                    <hr className={styles.divider} />
                                </Fragment>
                            )
                        }            
                    </div>
                 </>
                }
            </div>
        </section>
    )
};

export default Definition;

