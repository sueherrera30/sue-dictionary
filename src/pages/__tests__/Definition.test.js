import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import * as router from 'react-router'

import Definition from '../Definition';


const favoritesMock = ['squat', 'snatch', 'clean', 'thruster'];
const navigate = jest.fn()
const location = jest.fn()

afterEach(() => {
    cleanup();
});

beforeEach(() => {
    jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
    jest.spyOn(router, 'useLocation').mockImplementation(() => location)
})

test('<Definition />', () => {
    render(<MemoryRouter><Definition favorites={favoritesMock} /></MemoryRouter>, );
    expect(screen.container).toMatchSnapshot();
});

test('Back to home', async () => {
    render(<MemoryRouter><Definition favorites={favoritesMock} /></MemoryRouter>, );
    const backButton = screen.getByTestId('back-button');
    await fireEvent.click(backButton);
    expect(navigate).toHaveBeenCalledWith('/')
});

test('Back to favorites', async () => {
    render(<MemoryRouter><Definition favorites={favoritesMock} /></MemoryRouter>, );
    const favButton = screen.getByTestId('favorites-button');
    await fireEvent.click(favButton);
    expect(navigate).toHaveBeenCalledWith('/favorites');
});

test('verify is a word saved', () => {
   render(<MemoryRouter><Definition favorites={favoritesMock} /></MemoryRouter>,);
   const word = screen.queryByTestId('word').outerHTML;
   console.log(word);
  // expect(location.pathname).toBe(`/definition/${word}`);
});