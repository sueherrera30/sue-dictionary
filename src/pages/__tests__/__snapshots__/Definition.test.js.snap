// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Definition /> 1`] = `undefined`;
